File: leetcode/medium/Q19_RemoveNthNodeFromEndofList.py
# Definition for singly-linked list.

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        if not head:
            return head
        nodelist = []
        tmp = head
        while tmp:
            nodelist.append(tmp)
            tmp = tmp.next
        if n > len(nodelist):
            n = len(nodelist)

        if n == len(nodelist):
            head = head.next
        elif n == 1:
            nodelist[-n - 1].next = None
        else:
            nodelist[-n - 1].next = nodelist[-n + 1]
        return head

File: src/algorithm1/twopointer/Q19_RemoveNthNodeFromEndofList.java
package algorithm1.twopointer;

public class Q19_RemoveNthNodeFromEndofList {
     public class ListNode {
         int val;
         ListNode next;
         ListNode() {}
         ListNode(int val) { this.val = val; }
         ListNode(int val, ListNode next) { this.val = val; this.next = next; }
     }
    class Solution {
        public ListNode removeNthFromEnd(ListNode head, int n) {
            ListNode left = head;
            ListNode right = head;
            int length = 0;

            while (right != null){
                if(length > n) {
                    left = left.next;
                }
                right = right.next;
                length += 1;
            }
            if(length <= n){
                head = head.next;
            }else{
                ListNode temp = left.next;
                left.next = temp == null ? null : temp.next;
            }
            return head;
        }
    }
}
